File: _archived/notifications-nodejs/src/integrations/webhooks/normalizers/sendgrid.normalizer.ts
import { NormalizedEventType } from "../../../models/notification-event.model";

export interface SendGridEventItem {
  event: string;
  sg_message_id?: string;
  timestamp?: number;
  email?: string;
  reason?: string;
  status?: string;
  type?: string;
  url?: string;
  [key: string]: unknown;
}

export interface NormalizedProviderEvent {
  rawEventType: string;
  normalizedEventType: NormalizedEventType;
  providerMessageId: string | null;
  eventTimestamp: Date;
  recipientEmail: string | null;
  metadata: Record<string, unknown>;
}

const SENDGRID_EVENT_MAP: Record<string, NormalizedEventType> = {
  processed: "accepted",
  deferred: "deferred",
  delivered: "delivered",
  bounce: "bounced",
  blocked: "rejected",
  dropped: "rejected",
  open: "opened",
  click: "clicked",
  spamreport: "complained",
  unsubscribe: "unsubscribed",
  group_unsubscribe: "unsubscribed",
  group_resubscribe: "accepted",
  machine_opened: "opened",
};

function extractSgMessageId(raw: SendGridEventItem): string | null {
  const id = raw.sg_message_id;
  if (!id) return null;
  // sg_message_id may have a suffix like ".filter0001p3mdm1-17502-5F4B5765-1.0"
  // The core ID is the first segment
  return id.split(".")[0] ?? id;
}

export function normalizeSendGridEvent(raw: SendGridEventItem): NormalizedProviderEvent {
  const rawEventType = raw.event || "unknown";
  const normalizedEventType: NormalizedEventType =
    SENDGRID_EVENT_MAP[rawEventType] ?? "failed";

  const ts = raw.timestamp ? new Date(raw.timestamp * 1000) : new Date();
  const { event, sg_message_id, timestamp, email, ...rest } = raw;

  return {
    rawEventType,
    normalizedEventType,
    providerMessageId: extractSgMessageId(raw),
    eventTimestamp: ts,
    recipientEmail: raw.email ?? null,
    metadata: rest,
  };
}

File: apps/services/notifications/Notifications.Infrastructure/Webhooks/Normalizers/SendGridNormalizer.cs
using System.Text.Json;

namespace Notifications.Infrastructure.Webhooks.Normalizers;

public class SendGridEventItem
{
    public string Event { get; set; } = string.Empty;
    public string? SgMessageId { get; set; }
    public long? Timestamp { get; set; }
    public string? Email { get; set; }
    public string? Reason { get; set; }
    public string? Status { get; set; }
    public string? Type { get; set; }
    public string? Url { get; set; }
}

public class NormalizedSendGridEvent
{
    public string RawEventType { get; set; } = string.Empty;
    public string NormalizedEventType { get; set; } = string.Empty;
    public string? ProviderMessageId { get; set; }
    public DateTime EventTimestamp { get; set; }
    public string? RecipientEmail { get; set; }
    public Dictionary<string, object> Metadata { get; set; } = new();
}

public static class SendGridNormalizer
{
    private static readonly Dictionary<string, string> EventMap = new()
    {
        ["processed"] = "accepted",
        ["deferred"] = "deferred",
        ["delivered"] = "delivered",
        ["bounce"] = "bounced",
        ["blocked"] = "rejected",
        ["dropped"] = "rejected",
        ["open"] = "opened",
        ["click"] = "clicked",
        ["spamreport"] = "complained",
        ["unsubscribe"] = "unsubscribed",
        ["group_unsubscribe"] = "unsubscribed",
        ["group_resubscribe"] = "accepted",
        ["machine_opened"] = "opened"
    };

    public static NormalizedSendGridEvent Normalize(SendGridEventItem raw)
    {
        var rawEventType = raw.Event ?? "unknown";
        var normalizedEventType = EventMap.GetValueOrDefault(rawEventType, "failed");
        var ts = raw.Timestamp.HasValue
            ? DateTimeOffset.FromUnixTimeSeconds(raw.Timestamp.Value).UtcDateTime
            : DateTime.UtcNow;

        string? messageId = null;
        if (!string.IsNullOrEmpty(raw.SgMessageId))
        {
            var dotIndex = raw.SgMessageId.IndexOf('.');
            messageId = dotIndex > 0 ? raw.SgMessageId[..dotIndex] : raw.SgMessageId;
        }

        return new NormalizedSendGridEvent
        {
            RawEventType = rawEventType,
            NormalizedEventType = normalizedEventType,
            ProviderMessageId = messageId,
            EventTimestamp = ts,
            RecipientEmail = raw.Email,
            Metadata = new Dictionary<string, object>()
        };
    }

    public static List<SendGridEventItem> ParseEvents(string rawBody)
    {
        try
        {
            var doc = JsonDocument.Parse(rawBody);
            if (doc.RootElement.ValueKind == JsonValueKind.Array)
            {
                return doc.RootElement.EnumerateArray().Select(ParseSingleEvent).ToList();
            }
            return new List<SendGridEventItem> { ParseSingleEvent(doc.RootElement) };
        }
        catch
        {
            return new List<SendGridEventItem>();
        }
    }

    private static SendGridEventItem ParseSingleEvent(JsonElement element)
    {
        return new SendGridEventItem
        {
            Event = element.TryGetProperty("event", out var e) ? e.GetString() ?? "" : "",
            SgMessageId = element.TryGetProperty("sg_message_id", out var m) ? m.GetString() : null,
            Timestamp = element.TryGetProperty("timestamp", out var t) && t.TryGetInt64(out var ts) ? ts : null,
            Email = element.TryGetProperty("email", out var em) ? em.GetString() : null,
            Reason = element.TryGetProperty("reason", out var r) ? r.GetString() : null,
        };
    }
}
